src/bin/str.rs

```diff
@@ -89,88 +132,92 @@ define_language! {
-                if 0 <= i && i < (a.len() as _) {
-                    let i = i as usize;
-                    Some(Constant::Str(a[i..i+1].into()))
+                if i < 0 || i > (a.len() as _) {
+                    Some(Constant::Str(SmallString::default()))
                 } else {
-                    Some(Constant::Str(SmallStr::default()))
+                    let i = i as usize;
+                    Some(Constant::Str(SmallString::from(a.get(i..i+1).unwrap_or(""))))
                 }
             }),
             Lang::SubStr([a, i, len]) => map!(v, a, i, len => {
-                let a = a.to_slice().unwrap();
-                let n = a.len() as i64;
-                let i = i.to_num().unwrap();
-                let len = len.to_num().unwrap();
-                if 0 <= i && i < n && len > 0 {
-                    let j = (i + len).min(n) as usize;
-                    let i = i as usize;
-                    Some(Constant::Str(a[i..j].into()))
-                } else {
-                    Some(Constant::Str(SmallStr::default()))
+                let a = a.to_smallstr().unwrap();
+                let n = a.len() as usize;
+                let i = i.to_num().unwrap() as usize;
+                let len = len.to_num().unwrap() as usize;
+                let j = (i + len).min(n) as usize;
+                match a.get(i..j) {
+                    Some(s) => Some(Constant::Str(SmallString::from(s))),
+                    None => Some(Constant::Str(SmallString::default()))
                 }
             }),
             Lang::FromInt(i) => map!(v, i => {
                 let i = i.to_num().unwrap();
-                Some(Constant::Str(i.to_string().into_bytes().into()))
+                Some(Constant::Str(SmallString::from(i.to_string())))
             }),
             Lang::ToInt(s) => map!(v, s => {
                 let s = s.to_str().unwrap();
                 let i = s.parse::<i64>().unwrap_or(-1);
                 Some(Constant::Num(i))
             }),
-            Lang::Len(s) => map!(v, s => {
-                let s = s.to_str().unwrap();
-                Some(Constant::Num(s.len() as _))
-            }),
+            Lang::Len(s) => {
+                map!(v, s => Some(Constant::Num(s.to_smallstr().unwrap().len() as i64)))
+            }
             Lang::IndexOf([s, t, i]) => map!(v, s, t, i => {
                 let s = s.to_str().unwrap();
                 let t = t.to_str().unwrap();
-                let i = i.to_num().unwrap();
-                if 0 <= i && i <= (s.len() as _) {
-                    let i = i as usize;
-                    match s[i..].find(t) {
-                        None => Some(Constant::Num(-1)),
-                        Some(j) => Some(Constant::Num(j as _))
-                    }
-                } else {
-                    Some(Constant::Num(-1))
+                let i = i.to_num().unwrap() as usize;
+                match s.get(i..s.len()).unwrap_or("").find(t) {
```

Comment:
i have the same question here as I had for `SubStr` but I think the `match` probably does the right thing here too? It will just be `None` if the bounds are not right?

Reply:
Yeah, if the bounds are not right, the `get()` will return `None`, and then `None.unwrap_or("")` will be `""`, and then you call `find` on that (which will also return `None`)